File: leetcode-js/148-sort-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var sortList = function(head) {
    if (!head || !head.next) {
        return head;
    }

    let fast = head.next.next;
    let slow = head;

    while (fast && fast.next) {
        fast = fast.next.next;
        slow = slow.next;
    }

    const head1 = sortList(slow.next);
    slow.next = null;
    const head2 = sortList(head);

    return mergeList(head1, head2);
};

function mergeList(head1, head2) {
    const dummy = new ListNode(0);
    let node = dummy;

    while (head1 && head2) {
        if (head1.val < head2.val) {
            node.next = head1;
            head1 = head1.next;
        } else {
            node.next = head2;
            head2 = head2.next;
        }
        node = node.next;
    }

    node.next = head1 ? head1 : head2;

    return dummy.next;
}

File: leetcode-java/148-sort-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode sortList(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }

        ListNode prev = null;
        ListNode slow = head;
        ListNode fast = head;

        while (fast != null && fast.next != null) {
            prev = slow;
            slow = slow.next;
            fast = fast.next.next;
        }

        prev.next = null;

        ListNode node1 = sortList(head);
        ListNode node2 = sortList(slow);

        return merge(node1, node2);
    }

    ListNode merge(ListNode node1, ListNode node2) {
        ListNode dummy = new ListNode(0);
        ListNode node = dummy;

        while (node1 != null && node2 != null) {
            if (node1.val < node2.val) {
                node.next = node1;
                node1 = node1.next;
            } else {
                node.next = node2;
                node2 = node2.next;
            }
            node = node.next;
        }

        if (node1 != null) {
            node.next = node1;
        } else if (node2 != null) {
            node.next = node2;
        }

        return dummy.next;
    }
}
